README.md

```diff
@@ -217,3 +217,61 @@ The resulting JSON structure derives the metric name and tag-map by decomposing
+    - A raw metric with a name that already ends with an aggregate qualifier, will only be aggregated with the same aggregation with the exception of "average" (since an average of an average is not mathematically robust). For example, if a raw metric is named `cpu_usage_idle_min` then only a min-aggregation is performed
+    - A configurable list of "counter suffixes" is used to further identify those raw metrics that should only be aggregated with a sum-aggregation. Examples of such suffixes are "reads", "writes", "bytes"
+    - The downsample process will then iterate through the retrieve raw data values, which were supplied in time ascending order, and incrementally aggregate the values. [Commons Math's description statistics](https://commons.apache.org/proper/commons-math/userguide/stat.html#a1.2_Descriptive_statistics) or similar could be used to perform those per-granularity aggregations.
+    - As each granularity interval of the raw data set is aggregated, an aggregated data row is upserted for that tenant, series-set, and normalized timestamp.
+        - As mentioned above, Cassandra doesn't have an "upsert", but `UPDATE` effectively acts as one. 
+        - It's important that the aggregate data row be `UPDATE`d rather than `INSERT`ed to accommodate late arrival handling, describe below
+        - The TTL used for the `UPDATE` will be determined by the retention duration configured for that granularity. For example, 5m granularity might configured with 14 days retention and 1h with 1 year retention.
+    - Upon completion of each `pending_downsample_sets` result row, that row is deleted from the table
+- Late arrivals of metrics into `data_row` are an interesting case to confirm are covered by the strategy above. There are two versions of late arrivals:
+    - No metrics at all showed up for a given tenant+series-set during a time slot. 
+        - That case is handled by the design above simply because the `pending_downsample_sets` row never would have been created. 
+        - When those metrics arrive later, the `pending_downsample_sets` row is created at that time and the downsample process will agreegate those as normal.
+    - If a subset of metrics showed up for a given tenant+series-set during a time slot
+        - The downsample processor won't know that there's a subset of the expected metrics; however, that's fine because all of the aggregations are mathematically correct for what is present. For example, the "count" aggregation would be accurate even if it's a lesser value than was expected by an end user
+        - When the remaining metrics arrive later, the `pending_downsample_sets` row is **recreated**
+        - The downsample processor will pick up on the pending row at the next scheduled time and *assuming the raw data for the time slot has not TTL'ed away*, the re-query of that time slot's raw data will now retrieve all expected metrics entries
```

Comment:
I wonder if we should just drop any delayed metrics that arrive close the time when the expected raw data would have ttl'd away?

Reply:
Interesting point. I'll include that in my further thinking.